src/input_common.h

```diff
@@ -12,91 +12,185 @@
+        kill_whole_line,
+        kill_inner_line,
+        kill_word,
+        kill_bigword,
+        backward_kill_word,
+        backward_kill_path_component,
+        backward_kill_bigword,
+        history_token_search_backward,
+        history_token_search_forward,
+        self_insert,
+        self_insert_notfirst,
+        transpose_chars,
+        transpose_words,
+        upcase_word,
+        downcase_word,
+        capitalize_word,
+        togglecase_char,
+        togglecase_selection,
+        execute,
+        beginning_of_buffer,
+        end_of_buffer,
+        repaint_mode,
+        repaint,
+        force_repaint,
+        up_line,
+        down_line,
+        suppress_autosuggestion,
+        accept_autosuggestion,
+        begin_selection,
+        swap_selection_start_stop,
+        end_selection,
+        kill_selection,
+        insert_line_under,
+        insert_line_over,
+        forward_jump,
+        backward_jump,
+        forward_jump_till,
+        backward_jump_till,
+        func_and,
+        func_or,
+        expand_abbr,
+        delete_or_exit,
+        exit,
+        cancel_commandline,
+        cancel,
+        undo,
+        redo,
+        begin_undo_group,
+        end_undo_group,
+        repeat_jump,
+        disable_mouse_tracking,
+        set_cursor,
+        insert_chars,
+        // NOTE: This one has to be last.
+        reverse_repeat_jump
+    };
+
+    /// Get the command this instance represents
+    id_t id() const { return id_; }
+
+    /// Get the data associated with set_cursor command.
+    const commandline_cursor_pos_t& get_cursor_pos() const {
+        assert(id_ == id_t::set_cursor && "Only valid for set_cursor");
+        return arg_.cursor_pos;
+    }
+
+    /// Get the data associated with insert_chars command.
+    const commandline_insertion_t& get_insertion() const {
+        assert(id_ == id_t::insert_chars && "Only valid for insert_chars");
+        return arg_.insertion;
+    }
+
+    /// Create a command without data.
+    /* implicit */ readline_cmd_t(readline_cmd_t::id_t id) : id_(id) {
+        assert(id_ != id_t::set_cursor && id_ != id_t::insert_chars &&
+               "Cannot create set_cursor or insert_chars commands with this constructor");
+    }
+
+    /// Create set_cursor command.
+    /* implicit */ readline_cmd_t(commandline_cursor_pos_t pos)
+        : id_(id_t::set_cursor), arg_(pos) {}
+
+    /// Create insert_chars command.
+    /* implicit */ readline_cmd_t(commandline_insertion_t insertion)
+        : id_(id_t::insert_chars), arg_(std::move(insertion)) {}
+
+    readline_cmd_t(const readline_cmd_t& rl);
+
+    readline_cmd_t(readline_cmd_t&& rl) noexcept;
+
+    readline_cmd_t& operator=(const readline_cmd_t& rl);
+
+    readline_cmd_t& operator=(readline_cmd_t&& rl) noexcept;
+
+    ~readline_cmd_t();
+
+   private:
+    id_t id_;
+
+    union arg_t {
+        /// Set if id_ is set_cursor
+        commandline_cursor_pos_t cursor_pos;
```

Comment:
Yeah, this is probably the best solution for now.
Just out of curiosity, here are some possible alternatives to the union:

1.

```cpp
class readline_cmd_arg_t {
	virtual ~readline_cmd_arg_t();
};
class readline_cmd_cursor_pos_arg_t : readline_cmd_arg_t {
	commandline_part_t part;
	long pos;
}
struct readline_cmd_t {
	id_t id_;
	std::unique_ptr<readline_cmd_arg_t> arg_;
}
```

2.

```cpp
class readline_cmd_t {
	virtual ~readline_cmd_arg_t();
};
class readline_cmd_beginning_of_line_t : readline_cmd_t {};
class readline_cmd_cmd_cursor_pos_t : readline_cmd_t {
	commandline_part_t part;
	long pos;
};
```

The second is closest to Rust enums but it probably does not make for a good for developer experience in C++.